flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/ScheduledFutureAdapter.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.concurrent;
+
+import javax.annotation.Nonnull;
+
+import java.util.concurrent.Delayed;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Adapter from {@link Future} to {@link ScheduledFuture}. This enriches the basic future with scheduling information.
+ * @param <V> value type of the future.
+ */
+public class ScheduledFutureAdapter<V> implements ScheduledFuture<V> {
```

Comment:
I think, in general, this is harly possible because it means that we need to consider `delegate` as well in the `compareTo` method as a tie-breaker to match it with `equals` because I assume an equality check should not ignore the delegate future. However, I see no way to also create a fixed, collision free ordering on objects of type `Future`. So this might be a place where the `not required` applies?

Reply:
I will introduce an UUID field, just for tie breakers.

Reply:
I was thinking about introducing a UUID as tie-breaker, but this will also not help in the general case, because `#compareTo(...)` goe against the `Delayed` interface.

Reply:
Nicely solved with the sequence number.